### PR TITLE
MSI engineering system

### PR DESCRIPTION
**Pull request by danieljurek**
Builds MSI, in release cases also builds and signs MSI with signed .exe. 

Depends on merging of a azure-sdk-build-tools PR: https://dev.azure.com/azure-sdk/internal/_git/azure-sdk-build-tools/pullrequest/588 ... Once that's merged I can clean up the TODOs around setting the right version. 

MSI supports a subset of SemVer as its versioning scheme. As such it does cannot present "beta" versions or handle other non-release scenarios like PRs in the same way that we can version the CLI. 

With @heaths' help I've applied an algorithm that has been used for shipping prerelease versions of MSI packages in the past. This involves compromising on the "patch" version numbers so that packages supersede properly. It's not un-common to have MSI packages with differing versions from the specific tools they contain. 

This logic lives in `Get-MsiVersion.ps1`: 

* If pre-release: version Patch * 100 + pre-release number 
* If not pre-release: version Patch + 1 * 100 

| CLI version | Calculated MSI version | 
| ----------- | ---------------------- |
| 0.4.0-beta.2 | 0.4.2 | 
| 0.4.0 | 0.4.100 | 
| 0.4.1-beta.1 | 0.4.101 | 
| 0.4.1-beta.2 | 0.4.102 | 
| 0.4.1 | 0.4.200 | 
| 1.0.0 | 1.0.100 | 

This ensures that released versions will always supersede previously released versions and the user can upgrade without having to uninstall the previous release.

### CODE REVIEW

File: cli/installer/windows/azd.wixproj [hunk 1]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (heaths) [inline cli/installer/windows/azd.wixproj:1]: Why do these files look new in this PR when I've already merged them?

File: cli/installer/windows/local-build.ps1 [hunk 2]
```diff
@@ -0,0 +1,12 @@
+param(
+    $Version = "0.0.1"
+)
+
+$currentLocation = Get-Location 
+
+try { 
+    Set-Location $PSScriptRoot
```
Comment (heaths) [inline cli/installer/windows/local-build.ps1:8]:
```suggestion
    Push-Location $PSScriptRoot
```

File: cli/installer/windows/local-build.ps1 [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+    dotnet msbuild /p:ProductVersion=$Version
+} finally { 
+    Set-Location $currentLocation
```
Comment (heaths) [inline cli/installer/windows/local-build.ps1:11]:
```suggestion
    Pop-Location
```

File: cli/installer/windows/azd.wixproj (hunk 1 above)
Comment (danieljurek) [reply cli/installer/windows/azd.wixproj]: Needed to do a rebase

File: cli/installer/windows/local-build.ps1 (hunk 2 above)
Comment (danieljurek) [reply cli/installer/windows/local-build.ps1:8]: I've generally avoided `Push` and `Pop` locations. I don't think they're going to cause problems in this case but if future iterations of this script include other scripts (or user defined script blocks) then it's possible that nested calls to `Push-Location` would not run the corresponding `Pop-Location` as expected and that would mean that our `Pop-Location` puts the user back in an unexpected state.

Comment (heaths) [reply cli/installer/windows/local-build.ps1:8]: Fair enough, but I would argue that's a bug on the callee script that should be fixed as well.

File: eng/pipelines/release-cli.yml [hunk 4]
```diff
@@ -404,7 +405,7 @@ stages:
               pwsh: true
               targetType: filePath
               filePath: cli/installer/windows/test-win-msi.ps1
-              arguments: $(AdditionalTestParameters)
+              arguments: -PerMachine:$$(PerMachine)
```
Comment (heaths) [inline eng/pipelines/release-cli.yml:408]: Was the double dollar intentional? Not familiar with that syntax.

Comment (danieljurek) [reply eng/pipelines/release-cli.yml:408]: This syntax comes from the interplay between PowerShell and DevOps' macro syntax.

It handles encoding for the cmdlet switch. The `$(PerMachine)` is rewritten as the literal string `true` or `false` and the expression becomes `-PerMachine:$false` and PowerShell can parse. 

If we used a single dollar sign (i.e. `-PerMachine:$(PerMachine)`) the result would translate to `-PerMachine:false` and you'd get an argument error for passing a string to `-PerMachine` instead of a `SwitchParameter` 

We could also use `1` or `0` in the matrix value and avoid double dollar sign.

File: eng/pipelines/release-cli.yml [hunk 5]
```diff
@@ -338,33 +326,109 @@ stages:
...
+            Copy-Item win/ signed-win/ -Recurse
+            Copy-Item win/azd-windows-amd64.exe cli/azd/azd.exe
+          displayName: Copy signing outputs for publishing and MSI build
+          condition: always()
+
+        - task: MSBuild@1
+          displayName: Build release MSI
+          inputs:
+            solution: cli/installer/windows/azd.wixproj
+            msbuildArguments: /p:RunWixToolsOutOfProc=true
```
Comment (heaths) [inline eng/pipelines/release-cli.yml:412]: /p:Configuration=Release

MSIs build Debug by default. Unless Configuration is being set as an env var somewhere prior.

File: eng/pipelines/release-cli.yml [hunk 6]
```diff
@@ -764,17 +876,29 @@ stages:
...
+              PowerShell install
+
               ``````
               powershell -c "Set-ExecutionPolicy Bypass Process; irm '$urlBase/uninstall-azd.ps1' > uninstall-azd.ps1; ./uninstall-azd.ps1;"
               powershell -c "Set-ExecutionPolicy Bypass Process; irm '$urlBase/install-azd.ps1' > install-azd.ps1; ./install-azd.ps1 -BaseUrl '$urlBase' -Version '' -Verbose;"
               ``````
 
+              MSI install
+              ``````
+              powershell -c "irm '$msiReleaseUrl' -OutFile azd-windows-amd64.msi; msiexec /i azd-windows-amd64.msi /qn"
```
Comment (heaths) [inline eng/pipelines/release-cli.yml:889]: Shouldn't the script i.e. _install-azd.ps1_ just do this itself? Seems more trouble to bifurcate instructions for the user. Packaging it as an MSI makes it installable from more locations e.g., chocolatey, scoop, and winget that we will do in another PR.

**Review by heaths — APPROVED** [review]

File: eng/pipelines/release-cli.yml (hunk 6 above)
Comment (danieljurek) [reply eng/pipelines/release-cli.yml:889]: We'll do that in a separate PR. I want to have an official release of the MSI before we put that into the install script. Otherwise we'd break the install script in a pretty fundamental way if it tried to download an MSI that didn't exist.
